### PR TITLE
Adds: RegistrationLifecycle

### PR DESCRIPTION
**Pull request by huifer**
This PR is used to solve this issue[#999](https://github.com/spring-cloud/spring-cloud-commons/issues/999)

### CODE REVIEW

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/serviceregistry/AbstractAutoServiceRegistration.java [hunk 1]
```diff
@@ -130,16 +135,45 @@ public void start() {
...
+								.postProcessAfterStartRegisterManagement(getManagementRegistration()));
+
 			}
 			this.context.publishEvent(new InstanceRegisteredEvent<>(this, getConfiguration()));
 			this.running.compareAndSet(false, true);
 		}
 
 	}
 
+	private List<RegistrationManagementLifecycle> getRegistrationManagementLifecycles() {
```
Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/serviceregistry/AbstractAutoServiceRegistration.java:161]: I would prefer these be passed in via the constructor rather be retrieved from the context in this case.

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]
Please add tests

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/serviceregistry/AbstractAutoServiceRegistration.java (hunk 1 above)
Comment (huifer) [reply spring-cloud-commons/src/main/java/org/springframework/cloud/client/serviceregistry/AbstractAutoServiceRegistration.java]: My intention is to get it directly from the container, not from a member variable.  However,I'm going to modify this part of the code

**Comment by huifer** [comment]
> Please add tests

Related use cases have been added

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
@huifer thanks for submitting this PR. In general, it looks good and seems like a good solution to the issue. However, there are some cosmetic issues; have added some comments - please address them.

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/serviceregistry/AbstractAutoServiceRegistration.java [hunk 2]
```diff
@@ -76,6 +82,32 @@ protected AbstractAutoServiceRegistration(ServiceRegistry<R> serviceRegistry,
...
+	}
+
+	protected AbstractAutoServiceRegistration(ServiceRegistry<R> serviceRegistry,
+			AutoServiceRegistrationProperties properties, List<RegistrationLifecycle<R>> registrationLifecycles) {
+		this.serviceRegistry = serviceRegistry;
+		this.properties = properties;
+		this.registrationLifecycles = registrationLifecycles;
+	}
+
+	public void addRegistrationManagementLifecycles(
```
Comment (OlgaMaciaszek) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/serviceregistry/AbstractAutoServiceRegistration.java:102]: Is this needed? Is there a reason all the `RegistrationLifecycle` objects can't be initialised in the constructor and the field storing them cannot be `final`?

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/serviceregistry/AbstractAutoServiceRegistration.java [hunk 3]
```diff
@@ -76,6 +82,32 @@ protected AbstractAutoServiceRegistration(ServiceRegistry<R> serviceRegistry,
...
+			List<RegistrationManagementLifecycle<R>> registrationManagementLifecycles) {
+		this.registrationManagementLifecycles.addAll(registrationManagementLifecycles);
+	}
+
+	public void addRegistrationLifecycles(List<RegistrationLifecycle<R>> registrationLifecycles) {
```
Comment (OlgaMaciaszek) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/serviceregistry/AbstractAutoServiceRegistration.java:107]: Is this needed? Is there a reason all the `RegistrationLifecycle` objects can't be initialised in the constructor and the field storing them cannot be `final`?

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/serviceregistry/RegistrationManagementLifecycle.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.client.serviceregistry;
+
+/**
+ * Service registration life cycle. This life cycle is only related to
+ * ManagementRegistration.
```
[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-commons/src/main/java/org/springframework/cloud/client/serviceregistry/RegistrationManagementLifecycle.java:21]: Please implement the change requested above.

File: spring-cloud-commons/src/test/java/org/springframework/cloud/client/serviceregistry/AbstractAutoServiceRegistrationRegistrationLifecycleTests.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+		}
+
+		@Override
+		protected boolean isEnabled() {
+			return true;
+		}
+
+	}
+
+	public static class TestRegistrationLifecycleServiceRegistration
```
[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-commons/src/test/java/org/springframework/cloud/client/serviceregistry/AbstractAutoServiceRegistrationRegistrationLifecycleTests.java:165]: Please implement the change requested above.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Hello @huifer - thanks for adding the changes. Some previous comments have not been addressed, however - please address them. Also have added some comments related to recent changes - please take a look.

**Comment by huifer** [comment]
@OlgaMaciaszek  The related operations have been completed

**Comment by OlgaMaciaszek** [comment]
@huifer Thanks, but I still see some pending comments. Please address them.

**Comment by huifer** [comment]
If you can, please mark the unmodified for me



| |
***@***.***
|
|
***@***.***
|




---- Replied Message ----
| From | Olga ***@***.***> |
| Date | 05/25/2022 19:15 |
| To | ***@***.***> |
| Cc | Zen ***@***.******@***.***> |
| Subject | Re: [spring-cloud/spring-cloud-commons] Adds: RegistrationLifecycle (PR #1044) |

@huifer Thanks, but I still see some pending comments. Please address them.

**Comment by huifer** [comment]
@OlgaMaciaszek   I think the repair has been completed. If there is any incomplete repair, please help to point it out. Thank you

**Review by OlgaMaciaszek — APPROVED** [review]
Thanks for fixing these @huifer. LGTM.

**Pull request merged**